### PR TITLE
DO-46 <small changes in everything>

### PR DESCRIPTION
**Pull request by Katashynskyi**
Check it out. @ViktorDidkovskyi

### CODE REVIEW

**Comment by Katashynskyi** [comment]
fixed basemodel (working) & xgboost (working), lightgbm (in progress). How can I share mlflow runs, through csv only?

File: Testing/empty.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+doc = nlp("I have a banana")
+
+# Access the vector via the token.vector attribute
+print(doc[3].vector)
+
+# Print the vector representation of each token
+for token in doc:
+    print(token.text)#, token.vector)
+    print(token.shape)
+    print(len(token))
```
Comment (ViktorDidkovskyi) [inline Testing/empty.py]: Should not be a part of PR,

File: src/utils/features_preprocessing.py [hunk 2]
```diff
@@ -223,6 +226,35 @@ def stack(self, tfidf, norm_indirect_features):
...
+#     def __init__(self, nlp):
+#         self.nlp = nlp
+#         self.dim = 300
+#
+#     def fit(self, X, y):
+#         return self
+#
+#     def transform(self, X):
+#         return [self.nlp(text).vector for text in X]
+
```
Comment (ViktorDidkovskyi) [inline src/utils/features_preprocessing.py]: should be removed

File: src/utils/features_preprocessing.py [hunk 3]
```diff
@@ -223,6 +226,35 @@ def stack(self, tfidf, norm_indirect_features):
...
+    def fit(self, X, y=None):
+        return self
+
+    def transform(self, X):
+        docs = [self.nlp(text) for text in X]
+        outputs = [doc.cats.values() for doc in docs]
+        outputs = [list(x) + [0] * (self.output_dim - len(x)) for x in outputs]
+        return outputs
+
```
Comment (ViktorDidkovskyi) [inline src/utils/features_preprocessing.py]: better to have in seperate logical files

File: src/utils/features_preprocessing.py [hunk 4]
```diff
@@ -223,6 +226,35 @@ def stack(self, tfidf, norm_indirect_features):
         return hstack((tfidf, norm_indirect_features))
```
Comment (ViktorDidkovskyi) [inline src/utils/features_preprocessing.py]: with open("../../data/tfidf.pickle", 'rb') as file:
            return pickle.load(file)
            
            Path - should be a parameter,
            better to have something like: self._tfidf = pickle.load(file)
            Thinks maybe you should remove load, and save if you are not using it in any place

File: src/utils/features_preprocessing.py [hunk 5]
```diff
@@ -235,15 +267,15 @@ class Preprocessor:
...
-        self._tfidf = TfidfVectorizer(max_df=max_df, min_df=min_df,ngram_range=ngram_range,max_features=max_feat,norm=norm)
+    def __init__(self, vectorizer_type=None, scaler=MinMaxScaler()):
+        if vectorizer_type == 'tfidf':
+            self.vectorizer_type = 'tfidf'
+        elif vectorizer_type == 'spacy':
+            self.vectorizer_type = 'spacy'
+        else:
+            raise ValueError("Invalid embedding type: choose 'tfidf' or 'spacy'")
+        self.scaler = scaler
+        self._tfidf = CustomTfidf()
```
Comment (ViktorDidkovskyi) [inline src/utils/features_preprocessing.py]: Thinks how to improve if statement, now it looks bad
Also, you need to change the logic for self._tfidf = CustomTfidf()
for example, vectorizer_type = ' spacy'
your if statement will pass and   self._tfidf = CustomTfidf()- will also work. But it is not spacy,

File: src/utils/features_preprocessing.py [hunk 6]
```diff
@@ -259,7 +291,10 @@ def fit(self, x, y=None):
         --------
         self
         """
-        self._tfidf.fit(x)
+        if self.vectorizer_type == 'tfidf':
+            self._tfidf.fit(x)
+        elif self.vectorizer_type == 'spacy':
+            pass
```
Comment (ViktorDidkovskyi) [inline src/utils/features_preprocessing.py]: better to have one self._vectorizer.fit
instead if

File: src/utils/features_preprocessing.py [hunk 7]
```diff
@@ -277,19 +312,24 @@ def transform(self, x, y=None):
         csr_matrix
             The sparse matrix of TF-IDF features.
         """
-        tfidf = self._tfidf.transform(x)
+        if self.vectorizer_type == 'tfidf':
+            tfidf = self._tfidf.transform(x)
```
Comment (ViktorDidkovskyi) [inline src/utils/features_preprocessing.py]: better to have one self._vectorizer.fit
instead if

File: src/utils/features_preprocessing.py [hunk 8]
```diff
@@ -277,19 +312,24 @@ def transform(self, x, y=None):
...
+            normalize = AddingFeatures(self.scaler).normalize(additional_features)
 
+            # AddingFeatures stack
+            stack = AddingFeatures().stack(tfidf, normalize)
+            return stack  # -> csr_matrix
+        elif self.vectorizer_type == 'spacy':
+            import spacy
+            nlp = spacy.load("en_core_web_md")
+            vectorizer_type = SpacyVectorTransformer(nlp=nlp)
+            return vectorizer_type
```
Comment (ViktorDidkovskyi) [inline src/utils/features_preprocessing.py]: you lost  AddingFeatures()

File: src/utils/features_preprocessing.py [hunk 9]
```diff
@@ -307,20 +347,25 @@ def fit_transform(self, x, y=None):
...
+            normalize = AddingFeatures().normalize(additional_features)
+            # print(normalize)
+            # AddingFeatures stack
+            stack = AddingFeatures().stack(tfidf, normalize)
+            return stack  # -> csr_matrix
+        elif self.vectorizer_type == 'spacy':
+            import spacy
+            nlp = spacy.load("en_core_web_md")
+            vectorizer_type = SpacyVectorTransformer(nlp=nlp)
+            return vectorizer_type
```
Comment (ViktorDidkovskyi) [inline src/utils/features_preprocessing.py]: duplicated code
maybe can have additional functions?

File: src/utils/features_preprocessing.py [hunk 10]
```diff
@@ -335,21 +380,24 @@ def fit_transform(self, x, y=None):
...
-    print(f"stack: {type(stack)}")
     # Preprocessor test
-    p = Preprocessor().fit_transform(train_X)
-    p = Preprocessor().fit(train_X).transform(train_X)
-    print(p)
+    p = Preprocessor(vectorizer_type='spacy').fit_transform(train_X)
+    # p = Preprocessor().fit(train_X).transform(train_X)
+    print(p.nlp) #  spacy.lang.en.English object at 0x00000264AEE66AC0
+    print(p.dim) #  300
+    print(p.__class__) #  class '__main__.SpacyVectorTransformer'
```
Comment (ViktorDidkovskyi) [inline src/utils/features_preprocessing.py]: maybe this should be test? or let's remove commented code

**Review by ViktorDidkovskyi — CHANGES_REQUESTED** [review]
Good work, some suggestions:
use [black](https://github.com/psf/black) + [sor](https://pypi.org/project/isort/) to make code prertty

Best practises PR 
Great to have a clear description what you have done, what is the purpose, etc
PR- should have working code with "green" test, without questions, without not working functions. For status, you have a jira
"small changes in everything" - says nothing about the context which should be checked 
If it is finish version of project, good practise to have a Read me or comments how to run the code
Good practise to have a small PR, easier to review

File: src/utils/features_preprocessing.py (hunk 10 above)
Comment (Katashynskyi) [reply src/utils/features_preprocessing.py]: test like separate file with unittest or pytest ?

**Comment by Katashynskyi** [comment]
@ViktorDidkovskyi 
Resolved all conversations, added spacy vectorizer, separated tfidf, addings features and spacy vectorizers from preprocessor.py. Everything functioning. 

Next, I'll try to add cupy (CUDA for numpy) to the project, 

The spacy too slow (more then 30m to 1 optimization step for medium model), I'll try to add easier vectorizers like word2vec, fasttext, glove etc.

Then autotests.

**Pull request merged**